tower/src/builder/mod.rs

```diff
@@ -0,0 +1,194 @@
+/// ```
+///
+/// A `Service` stack with a single layer:
+///
+/// ```
+/// # extern crate tower;
+/// # extern crate tower_in_flight_limit;
+/// # extern crate futures;
+/// # extern crate void;
+/// # use void::Void;
+/// # use tower::Service;
+/// # use tower::builder::ServiceBuilder;
+/// # use tower_in_flight_limit::InFlightLimitLayer;
+/// # use futures::{Poll, future::{self, FutureResult}};
+/// # #[derive(Debug)]
+/// # struct MyService;
+/// # impl Service<()> for MyService {
+/// #    type Response = ();
+/// #    type Error = Void;
+/// #    type Future = FutureResult<Self::Response, Self::Error>;
+/// #    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+/// #        Ok(().into())
+/// #    }
+/// #    fn call(&mut self, _: ()) -> Self::Future {
+/// #        future::ok(())
+/// #    }
+/// # }
+/// ServiceBuilder::new()
+///     .layer(InFlightLimitLayer::new(5))
+///     .build_service(MyService);
+/// ```
+///
+/// A `Service` stack with _multiple_ layers that contain rate limiting, in-flight request limits,
+/// and a channel-backed, clonable `Service`:
+///
+/// ```
+/// # extern crate tower;
+/// # extern crate tower_in_flight_limit;
+/// # extern crate tower_buffer;
+/// # extern crate tower_rate_limit;
+/// # extern crate futures;
+/// # extern crate void;
+/// # use void::Void;
+/// # use tower::Service;
+/// # use tower::builder::ServiceBuilder;
+/// # use tower_in_flight_limit::InFlightLimitLayer;
+/// # use tower_buffer::BufferLayer;
+/// # use tower_rate_limit::RateLimitLayer;
+/// # use std::time::Duration;
+/// # use futures::{Poll, future::{self, FutureResult}};
+/// # #[derive(Debug)]
+/// # struct MyService;
+/// # impl Service<()> for MyService {
+/// #    type Response = ();
+/// #    type Error = Void;
+/// #    type Future = FutureResult<Self::Response, Self::Error>;
+/// #    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+/// #        Ok(().into())
+/// #    }
+/// #    fn call(&mut self, _: ()) -> Self::Future {
+/// #        future::ok(())
+/// #    }
+/// # }
+/// ServiceBuilder::new()
+///     .layer(BufferLayer::new(5))
+///     .layer(InFlightLimitLayer::new(5))
+///     .layer(RateLimitLayer::new(5, Duration::from_secs(1)))
+///     .build_service(MyService);
+/// ```
+#[derive(Debug)]
+pub struct ServiceBuilder<L> {
+    layer: L,
+}
+
+impl ServiceBuilder<Identity> {
+    /// Create a new `ServiceBuilder` from a `MakeService`.
+    pub fn new() -> Self {
+        ServiceBuilder {
+            layer: Identity::new(),
+        }
+    }
+}
+
+impl<L> ServiceBuilder<L> {
+    /// Layer a new layer `T` onto the `ServiceBuilder`.
+    pub fn layer<T>(self, layer: T) -> ServiceBuilder<Chain<T, L>> {
+        ServiceBuilder {
+            layer: Chain::new(layer, self.layer),
+        }
+    }
+
+    /// Create a `ServiceBuilderMaker` from the composed middleware and transport.
+    pub fn build_maker<M, S, Target, Request>(self, maker: M) -> LayeredMakeService<M, L, Request>
```

Comment:
I was told that i fI wanted to bikeshed I should bikeshed now...  Can we improve on the names `build_maker` and `build_service`? Specifically, we don't refer to "makers" anywhere else.

`build_make_service` is a little more verbose but slightly clearer imo.

Or, we could call these different things entirely, e.g. `build()` and `with_service()`?

Reply:
Im totally fine with calling it `build_make_service` but I don't quite follow what your other suggestion was? Are you suggesting `build_with_service`? or change it from build to `with_service` to complete the "builder"?

Reply:
I'm fine w/ `build_make_service() -> impl MakeService` as it matches `build_service() -> impl Service`.

I'm not sure what `build()` vs. `with_service()` would be.

Reply:
I was suggesting the latter, though i think `build_with_service` also sounds fine.

Note, though, that the maker variant also "builds a service", though that service is one that happens to build services (MakeService is itself a Service)...  So we may want to find naming that relates more to the abstract/concrete nature of the service rather than whether it is a service or not...

Reply:
Though `MakeService` is a service, we use the trait alias to tell people that it isn't just a regular service, so I feel naming it build make service is clear to tell people that there is a difference. I'm not sure how else we could abstractly convey that. 

With the `build_with_service`, would then it also have to be `build_with_make_service` which seems a bit long.